### PR TITLE
sql: multi-region tenant region management tightening 

### PR DESCRIPTION
**Pull request by ajwerner**
### sql: ensure multi-region tenants only see their regions

A multi-region tenant (a secondary tenant with a multi-region
system database), should only see the regions in its system
database. This patch makes that so.

In order to achieve that goal, the patch introduces an interface
in the sql layer (`RegionsProvider`) and an implementation in
a new sql sub-package `regions`. It adds the interface to the
definition of the descs.Txn, for use in the context of sql
transactions.

### sql: check if region is in use before dropping from system db

We cannot remove a region from a tenant if that region is in
use in any database in that tenant. This commit adds the logic
to perform that check.

Epic: CRDB-2526

Informs #98530

Release note: None

### CODE REVIEW

File: pkg/sql/planner.go [hunk 1]
```diff
@@ -626,6 +627,14 @@ func (p *planner) InternalSQLTxn() descs.Txn {
 	return &p.internalSQLTxn
 }
 
+func (p *planner) regionsProvider() *regions.Provider {
+	if txn := p.InternalSQLTxn(); txn != nil {
+		_ = txn.Regions() // force initialization
```
Comment (chengxiong-ruan) [inline pkg/sql/planner.go:632]: nit, why don't we just `return txn.Regions()` ?

**Review by chengxiong-ruan — APPROVED** [review]
LGTM!

File: pkg/sql/planner.go (hunk 1 above)
Comment (ajwerner) [reply pkg/sql/planner.go:632]: It doesn't have the concrete type. We'd need to do a type assertion.

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Pull request merged**